parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriterError.java

```diff
@@ -0,0 +1,193 @@
+      List<PhoneBookWriter.PhoneNumber> phoneNumbers;
+      if (RANDOM.nextDouble() < .1) {
+        phoneNumbers = null;
+      } else {
+        int n = RANDOM.nextInt(4);
+        phoneNumbers = new ArrayList<>(n);
+        for (int i = 0; i < n; ++i) {
+          String kind = RANDOM.nextDouble() < .1 ? null : "kind" + RANDOM.nextInt(5);
+          phoneNumbers.add(new PhoneBookWriter.PhoneNumber(RANDOM.nextInt(), kind));
+        }
+      }
+      String name = RANDOM.nextDouble() < .1 ? null : "name" + RANDOM.nextLong();
+      PhoneBookWriter.User user = new PhoneBookWriter.User(RANDOM.nextLong(), name, phoneNumbers, location);
+      return PhoneBookWriter.groupFromUser(user);
+    }
+
+    private static TrackingByteBufferAllocator createAllocator(final int oomAt) {
+      return TrackingByteBufferAllocator.wrap(new DirectByteBufferAllocator() {
+        private int counter = 0;
+
+        @Override
+        public ByteBuffer allocate(int size) {
+          if (++counter >= oomAt) {
+            Assert.assertEquals(
+                "There should not be any additional allocations after an OOM", oomAt, counter);
+            throw new OutOfMemoryError("Artificial OOM to fail write");
+          }
+          return super.allocate(size);
+        }
+
+        @Override
+        public void release(ByteBuffer b) {
+          CleanUtil.cleanDirectBuffer(b);
+
+          // It seems, if the size of the buffers are small, the related memory space is not given back to the
+          // OS, so writing to them after release does not cause any identifiable issue. Therefore, we
+          // try to explicitly zero the address, so the jvm crashes for a subsequent access.
+          try {
+            if (BUFFER_ADDRESS != null) {
+              BUFFER_ADDRESS.setLong(b, 0L);
+            }
+          } catch (IllegalAccessException e) {
+            throw new RuntimeException("Unable to zero direct ByteBuffer address", e);
+          }
+        }
+      });
+    }
+
+    public static void main(String[] args) throws Throwable {
+      CompressionCodecName[] codecs = {
+        CompressionCodecName.UNCOMPRESSED,
+        CompressionCodecName.GZIP,
+        CompressionCodecName.SNAPPY,
+        CompressionCodecName.ZSTD,
+        CompressionCodecName.LZ4_RAW
+      };
+      for (int cycle = 0; cycle < 50; ++cycle) {
+        try (TrackingByteBufferAllocator allocator = createAllocator(RANDOM.nextInt(100) + 1);
+            ParquetWriter<Group> writer = ExampleParquetWriter.builder(
+                    new LocalOutputFile(Paths.get(args[0])))
+                .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
+                .withType(PhoneBookWriter.getSchema())
+                .withAllocator(allocator)
+                .withCodecFactory(CodecFactory.createDirectCodecFactory(
+                    new Configuration(), allocator, ParquetProperties.DEFAULT_PAGE_SIZE))
+                .withCompressionCodec(codecs[RANDOM.nextInt(codecs.length)])
```

Comment:
Why do we need to change the codec?

Reply:
I am using direct codec factory here with different codecs which also have direct memory allocated. I wanted to be sure that those are not failing in this scenario either.

Reply:
How about adding a comment to explicitly state this?